### ISSUE TITLE
What is the correct behaviour of dingo Router when adding named routes?

### ISSUE THREAD

**Issue by lightvision** [dingo/api#568]
When I add a named route, not a resource route, to the api routes, the API_PREFIX is ignored, if used.

On the other hand addind a resource route will correctly use API_PREFIX.

What is is the intended behaviour here?

**Comment by jasonlewis** [comment]
Can you give me an example please?

**Comment by lightvision** [comment]
```
$apiRouter=app('api.router');

$apiRouter->version('v1', function ($apiRouter) {
            $apiRouter->group([
                'prefix' => 'admin',
                'namespace' => $this->adminNamespace
                    ], function ($apiRouter) {
                $apiRouter->resource('clients','ClientController');
            });
        });

$apiRouteName = 'foo-api-endpoint.name';
$apiRouteUri = '/foo-api-endpoint';

// here i have doubts
$apiRouter->group([
            'version'=>'v1'
        ], function($apiRouter) use($apiRouteName,$apiRouteUri){
            $apiRouter->get($apiRouteUri,[
                'as'=>$apiRouteName,
                'uses'=>  function (){
                return [];
                }
            ]);
        });

```
my configuration is set to use API_PREFIX='api'

the foo-api-endpoint.name named route doesn't contain the 'api' like this: 'api.foo-api-endpoint.name'. The other routes defined as resource all have the prefix set before the actual route name, as so: 'api.admin.clients.index', 'api.admin.clients.create'.

**Comment by lightvision** [comment]
@jasonlewis What I am trying to say is that if I specify a name for a named route and if I use API_PREFIX then all named routes (excepting the resourceful routes) doesn't automatically prefixed. 

Is that the intended behavior?

**Comment by jasonlewis** [comment]
Ah right, yeah that's intended behaviour because the naming for resource routes is automatic, so it generates a name based on the URI, etc.

If you want consistency you'd probably be best making some helper function that prefixes it for you.

Like this I suppose:

```php
function name($name)
{
    return env('API_PREFIX').'.'.$name;
}
```

Then just use that to prefix the names.

```php
$apiRouteName = name('foo-api-endpoint.name');
```

**Comment by lightvision** [comment]
I don't think that this should be the correct behavior.

In an old version I had to add the 'API_PREFIX' myself but this was consistent across naming an endpoint.

Now I have to add myself that prefix for routes that are not resourceful. And because naming is inconsistent this leads to inconsistent tests.

**Comment by lightvision** [comment]
Me again:

if I set my routes like this:
```
public function map(ApiRouter $apiRouter) {
        $apiRouter->version('v1',['prefix'=>'api'],function ($apiRouter) {
            $apiRouter->group([
                'prefix' => 'admin',
                'namespace' => $this->adminNamespace
                    ], function ($apiRouter) {
                require app_path('Api/v1/Http/admin_routes.php');
            });
        });
    }
```
where ```'prefix'``` has the same value as `API_PREFIX` my prefix is ignored for all routes, no mater if they are resourceful or not.

If I made a second group with prefix I will end up with resourceful routes having *api.api.rest.of.name* names.

I have no choices to be consistent when declaring routes.

**Comment by lightvision** [comment]
@jasonlewis Any chances to have the prefix added to all routes, regardless of their type, when API_PREFIX is used?

**Comment by jasonlewis** [comment]
For starters, you don't need to redefine your prefix in the `version` group. It'll be added automatically for you. This saves duplicating your prefix should you change it in your `.env` file.

I'm not really sure what the problem here is. Manually naming a route shouldn't involve this package automatically adding something else to that name. That, to me, is unexpected.

Example.

```php
$api->version('v1', function ($api) {
    $api->get('/', ['as' => 'index', function () {
        return ['foo' => 'bar'];
    });
});
```

My route is named `index`, checking it with `api:routes` shows the name as `index`. That's expected. If it prefixed it with my `API_PREFIX` I would find that unexpected. That's not how I named it.

If *you* want that, then that's fine, but you'll need to add that naming functionality in yourself through a method such as the function I suggested earlier.

I don't see how that suggestion will duplicate the prefix on resource routes. Again, the example is:

```php
$api->version('v1', function ($api) {
    $api->get('/', ['as' => name('index'), function () {
        return ['foo' => 'bar'];
    });

    $api->resource('users', 'App\Http\Controllers\UserController');
});
```

My expected names are `api.index` for the `/` route, and `api.users.index` for the `/users` route. And, that's what I get.
